File: src/sites/myfreecams.cpp
// ─────────────────────────────────────────────────────────────────
// MyFreeCams site plugin — HTML scraping + model_id math
// Port from Python: share page scraping, campreview data attrs
// ─────────────────────────────────────────────────────────────────

#include "sites/myfreecams.h"
#include <regex>
#include <cstdint>

namespace sm
{

    REGISTER_SITE(MyFreeCams);

    MyFreeCams::MyFreeCams(const std::string &username)
        : SitePlugin(kSiteName, kSiteSlug, username)
    {
        sleepOnOffline_ = 10;
        sleepOnRateLimit_ = 60;
        maxConsecutiveErrors_ = 100;
    }

    std::string MyFreeCams::getWebsiteUrl() const
    {
        return "https://www.myfreecams.com/#" + username();
    }

    // Extract a data-attribute value from the campreview div
    std::string MyFreeCams::extractAttribute(const std::string &html,
                                             const std::string &attrName) const
    {
        // Search for data-cam-preview-{attrName}-value="..."
        std::string needle = "data-cam-preview-" + attrName + "-value=\"";
        auto pos = html.find(needle);
        if (pos == std::string::npos)
            return "";
        pos += needle.size();
        auto endPos = html.find('"', pos);
        if (endPos == std::string::npos)
            return "";
        return html.substr(pos, endPos - pos);
    }

    std::string MyFreeCams::buildPlaylistUrl() const
    {
        auto sidIt = attrs_.find("server-id");
        auto midIt = attrs_.find("model-id");
        auto wzobsIt = attrs_.find("is-wzobs");

        if (sidIt == attrs_.end() || midIt == attrs_.end())
            return "";

        std::string sid = sidIt->second;
        std::string midStr = midIt->second;

        if (sid.empty() || midStr.empty())
            return "";

        try
        {
            int64_t mid = 100000000 + std::stoll(midStr);
            std::string aPrefix = (wzobsIt != attrs_.end() && wzobsIt->second == "true") ? "a_" : "";

            return "https://previews.myfreecams.com/hls/NxServer/" + sid +
                   "/ngrp:mfc_" + aPrefix + std::to_string(mid) +
                   ".f4v_mobile_mhp1080_previewurl/playlist.m3u8";
        }
        catch (const std::exception &e)
        {
            return "";
        }
    }

    Status MyFreeCams::checkStatus()
    {
        // Step 1: Fetch the share page
        std::string shareUrl = "https://share.myfreecams.com/" + username();
        auto resp = http().get(shareUrl, 30);

        if (resp.isNotFound())
        {
            setLastError("Not found (404)", resp.statusCode);
            return Status::NotExist;
        }
        if (!resp.ok())
        {
            logger_->warn("HTTP {} for {}", resp.statusCode, username());
            setLastError("HTTP " + std::to_string(resp.statusCode), resp.statusCode);
            return Status::RateLimit;
        }

        const std::string &body = resp.body;
        setLastApiResponse(body.substr(0, 2048)); // Store first 2KB of HTML

        // Step 2: Look for tracking URL to verify model exists
        auto trackPos = body.find("https://www.myfreecams.com/php/tracking.php?");
        if (trackPos == std::string::npos)
            return Status::NotExist;

        // Verify model_id is in the tracking URL
        auto trackEnd = body.find('"', trackPos);
        if (trackEnd == std::string::npos)
            return Status::NotExist;

        std::string trackUrl = body.substr(trackPos, trackEnd - trackPos);
        if (trackUrl.find("model_id=") == std::string::npos)
            return Status::NotExist;

        // Step 3: Extract campreview data attributes
        auto camprevPos = body.find("class=\"campreview\"");
        if (camprevPos == std::string::npos)
        {
            // Also try alternative class name
            camprevPos = body.find("class='campreview'");
        }

        if (camprevPos == std::string::npos)
            return Status::Offline;

        // Extract a chunk around the campreview div
        size_t chunkStart = (camprevPos > 500) ? camprevPos - 500 : 0;
        size_t chunkEnd = std::min(camprevPos + 2000, body.size());
        std::string chunk = body.substr(chunkStart, chunkEnd - chunkStart);

        attrs_.clear();
        attrs_["server-id"] = extractAttribute(chunk, "server-id");
        attrs_["model-id"] = extractAttribute(chunk, "model-id");
        attrs_["is-wzobs"] = extractAttribute(chunk, "is-wzobs");

        if (attrs_["model-id"].empty())
            return Status::Offline;

        // Step 4: Build playlist URL and verify it
        std::string playlistUrl = buildPlaylistUrl();
        if (playlistUrl.empty())
            return Status::Offline;

        // Try to fetch the playlist to confirm it works
        auto playResp = http().get(playlistUrl, 30);
        if (playResp.ok())
        {
            videoUrl_ = selectResolution(playlistUrl);
            return videoUrl_.empty() ? Status::Private : Status::Public;
        }

        return Status::Private;
    }

    std::string MyFreeCams::getVideoUrl()
    {
        return videoUrl_;
    }

} // namespace sm

File: streamonitor/sites/myfreecams.py
import urllib.parse
from typing import Optional, Dict, Any, Tuple, List

import requests
from bs4 import BeautifulSoup
from streamonitor.bot import Bot
from streamonitor.enums import Status


class MyFreeCams(Bot):
    site: str = 'MyFreeCams'
    siteslug: str = 'MFC'

    def __init__(self, username: str) -> None:
        super().__init__(username)
        self.attrs: Dict[str, str] = {}
        self.videoUrl: Optional[str] = None
        self.url = self.getWebsiteURL()

    def get_site_color(self) -> Tuple[str, List[str]]:
        """Return the color scheme for this site."""
        return ("white", [])

    def getWebsiteURL(self) -> str:
        """Get the website URL for this streamer."""
        return f"https://www.myfreecams.com/#{self.username}"

    def getVideoUrl(self, refresh: bool = False) -> Optional[str]:
        """Get the video stream URL."""
        if not refresh:
            return self.videoUrl

        if 'data-cam-preview-model-id-value' not in self.attrs:
            return None

        try:
            sid = self.attrs.get('data-cam-preview-server-id-value', '')
            mid_str = self.attrs.get('data-cam-preview-model-id-value', '')
            
            if not sid or not mid_str:
                return None
                
            mid = 100000000 + int(mid_str)
            a = 'a_' if self.attrs.get('data-cam-preview-is-wzobs-value') == 'true' else ''
            
            playlist_url = f"https://previews.myfreecams.com/hls/NxServer/{sid}/ngrp:mfc_{a}{mid}.f4v_mobile_mhp1080_previewurl/playlist.m3u8"
            
            response = self.session.get(
                playlist_url,
                timeout=30,
                bucket='playlist'
            )
            
            if response.status_code != 200:
                return None
                
            return self.getWantedResolutionPlaylist(playlist_url)
            
        except (ValueError, KeyError) as e:
            self.logger.error(f"Error building video URL: {e}")
            return None
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Network error getting video URL: {e}")
            return None
        except Exception as e:
            self.logger.error(f"Unexpected error getting video URL: {e}")
            return None

    def getStatus(self) -> Status:
        """Check the current status of the stream."""
        try:
            response = self.session.get(
                f'https://share.myfreecams.com/{self.username}',
                timeout=30,
                bucket='status'
            )
            
            if response.status_code == 404:
                return Status.NOTEXIST
            elif response.status_code != 200:
                self.logger.warning(f"HTTP {response.status_code} for user {self.username}")
                return Status.UNKNOWN
                
            content = response.content
            
            # Look for tracking URL to verify model exists
            startpos = content.find(b'https://www.myfreecams.com/php/tracking.php?')
            if startpos == -1:
                return Status.NOTEXIST
                
            endpos = content.find(b'"', startpos)
            if endpos == -1:
                return Status.NOTEXIST
                
            url_bytes = content[startpos:endpos]
            try:
                url = urllib.parse.urlparse(url_bytes.decode('utf-8'))
                qs = urllib.parse.parse_qs(url.query)
                if 'model_id' not in qs:
                    return Status.NOTEXIST
            except Exception as e:
                self.logger.error(f"Error parsing tracking URL: {e}")
                return Status.NOTEXIST

            # Parse HTML for cam preview data
            try:
                doc = BeautifulSoup(content, 'html.parser')
                params = doc.find(class_='campreview')
                
                if params and params.attrs:
                    self.attrs = params.attrs
                    self.videoUrl = self.getVideoUrl(refresh=True)
                    
                    if self.videoUrl:
                        return Status.PUBLIC
                    else:
                        return Status.PRIVATE
                else:
                    return Status.OFFLINE
                    
            except Exception as e:
                self.logger.error(f"Error parsing HTML: {e}")
                return Status.ERROR
                
        except requests.exceptions.RequestException as e:
            self.logger.error(f"Network error checking status: {e}")
            return Status.ERROR
        except Exception as e:
            self.logger.error(f"Unexpected error: {e}")
            return Status.ERROR
    
    def isMobile(self) -> bool:
        """Check if this is a mobile broadcast."""
        return False


Bot.loaded_sites.add(MyFreeCams)
